### PR TITLE
Recipe to track skipped tests

### PR DESCRIPTION
**Pull request by MadinaB**
Could you please review following to check whether I am moving to right direction?
This is related to issue#27

### CODE REVIEW

**Comment by MadinaB** [comment]
Hello, @ahal!

Thank you for your explanations. Now it is more clear which kind of data should be returned. Could you please check it for now?

screen shot 2018-11-03 at 11 41 07 pm
(Due to long returned values it does not seem neat in default table format)

screen shot 2018-11-03 at 11 41 27 pm

**Comment by ahal** [comment]
Thanks, looks better!

Some notes:
1. You can just display the "fullname" of the suite rather than the whole config object
2. Looks like there's 1 really long test name that's messing up the table formatting. But that is a separate issue from this..
3. This is way too much data to be useful, so I think we need to limit it somehow. I see few options:
a) Require a platform to be specified on the command line (e.g linux64/opt)
b) Alternatively, require a suite on the command line and display platforms in the data
c) Restrict the data to the "mozilla-central" branch (actually we should make sure to do this anyway)

Anyway, this is getting close to a point where I'd be comfortable merging it and then improving iteratively.

File: adr/queries/skipped_tests.query [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from: unittest
+select:
+    - {aggregate: count}
+groupby:
+    - {value: result.test}
+    - {value: run.suite}
+where:
+    and:
```
Comment (ahal) [inline adr/queries/skipped_tests.query]: Please add

    - eq {repo.branch.name: "mozilla-central"}

That should cut down the amount of data a fair bit.

File: adr/recipes/skipped_tests.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from ..query import run_query
+from argparse import ArgumentParser
+
+
+def run(args, config):
+    parser = ArgumentParser()
+    parser.add_argument('--from', default='today-day', dest='from_date',
+                        help="Starting date to pull data from, defaults to a week ago")
+    parser.add_argument('--to', default='eod', dest='to_date',
+                        help="Ending date to pull data from, defaults to now")
```
Comment (ahal) [inline adr/recipes/skipped_tests.py]: For some reason if I run with the default values here I get no data back from ActiveData. Maybe this one needs to be "today"?

**Comment by MadinaB** [comment]
Thank you for the review! I will update that

**Comment by MadinaB** [comment]
There was a typo in "today-day" according to [here](https://github.com/mozilla/ActiveData/blob/dev/docs/jx_time.md) it should be at least "today-1day", but still it did not return any results. As well as, "today-3day" did not return any results, but "today-10day" started to return something. However, now I see that  "today-1day" is still incorrect, we should get data from "today" to "eod" in ideal. But still nothing is returned with this update:
screen shot 2018-11-07 at 12 04 33 am

**Comment by MadinaB** [comment]
Setting repo.branch.name to "mozilla-central" did not shorten results for a lot:

screen shot 2018-11-07 at 12 06 31 am

But specifying test suite actually did that well:

screen shot 2018-11-07 at 12 07 08 am

If no run suite is specified, it gets data from all suites. 

- That all suites' names are hard coded and I do not like this moment. In [code_coverage_by_suite.py](https://github.com/mozilla/active-data-recipes/blob/master/adr/recipes/code_coverage_by_suite.py) all suites are also hard coded, but I expect that if test suites are going to change, this data will not be actual any more. One solution could be create new query for getting all suites present in ActiveData.

**Comment by MadinaB** [comment]
Platform option is provided, as well, but it is not required one:

screen shot 2018-11-07 at 12 07 39 am


screen shot 2018-11-07 at 12 08 38 am

So if needed platform can be specified, but if not data from all platforms is being taken.

- All platforms info is also hard coded.

**Comment by ahal** [comment]
> Setting repo.branch.name to "mozilla-central" did not shorten results for a lot:

Yeah, I wouldn't expect this to reduce the number of tests specificially, but it should greatly reduce the `counts`. So while the end table might not look much smaller, the amount of data being processed should be. The `count` field isn't actually that important (it basically just says how many times the test ran). We mostly want to know whether the test was skipped in the most recent data we have.

**Comment by ahal** [comment]
> However, I do not think this is problem of code since in this time interval it still does not return anything with yyyy-mm-dd format:

Yeah I suspect your code is fine. Sometimes ActiveData lags a bit behind the live data. Let's just leave this alone for now.

**Comment by ahal** [comment]
>     * That all suites' names are hard coded and I do not like this moment. In [code_coverage_by_suite.py](https://github.com/mozilla/active-data-recipes/blob/master/adr/recipes/code_coverage_by_suite.py) all suites are also hard coded, but I expect that if test suites are going to change, this data will not be actual any more. One solution could be create new query for getting all suites present in ActiveData.

I agree that I don't like the fact that suites are hard-coded. I also like your solution, having a simple query like that would be a good building block for many different recipes! However in this case I think you could achieve the same thing by not doing any filtering by suite. The problem here is that we would either need two separate queries (one that has the suite filter and one that doesn't) or we'd need to build the query dynamically (depending on whether or not a suite was passed in). This is a problem that ADR could try to solve more generally but I'm not really sure how.

I'll leave it up to you what you prefer (either use two separate queries, or create a new query that returns the list of suites).

**Comment by MadinaB** [comment]
Hi, @ahal! 

Hope you are doing well!

I am so excited about this part :) Actually, how do you think about having some web or command line interface which would generate the queries dynamically? I have been thinking about exactly this part and I am stick to an idea of creating 2 different queries since we are running queries after writing them into file some-file.query and each time rewriting the file would not make sense. Currently, we are having recipe.py and query.py and recipe-code.py and the-query.query files to be needed to have only one query executed. Actually, as I understand (please, correct me if I got that wrong) the query is simply a dict we send to an ActiveData (we use jsone.render and dict of values taken from argparse to replace variables before sending) and later we simply write some python code to get that being rendered as we want to see it. I am thinking that if we would want the query to be generated dynamically there are 2 cases to do it:
1. Often used queries. It does make sense to have them written in some file since we will execute them often.
2. Rarely used queries. It does not make sense to save them since they would take space, some available query names and do query list messy.
**For often used queries:** So currently we are holding path of execution for often used queries. We could improve that by providing user friendly interface like a website for generation of these queries and recipes for calling them. I view it as some questionnaire where according to user answers we would generate that query dict and write it to new-query.query. User could provide some python code telling how he would want the query to be printed and we would add that into new-query.py. We could force user to give some explanation for his recipe and add it [to the list of available recipes](https://github.com/mozilla/active-data-recipes/tree/aab3346d60a5bcb948909b0684541024937b7e48/app) we plan to complete soon.
As well, it would be possible to create db and authorization for the website so that each user could save his own queries.
**For rarely used queries:** It would make sense to create some shorter script which would simply get query draft adjusted to input args. In our current example, if suite is not provided jsone.render will throw `KeyError` and `jsone.shared.InterpreterError`. If this is one time need we could have draft queries and update them to fit our needs so that if some value is not provided, we delete that value in the dict and pass dict to  query_activedata(query, url).

**Comment by MadinaB** [comment]
Hi, @ahal! 

Hope you are doing well!

I am so sorry for late reply.

I am very excited about solving this part :) Actually, I view solution for this in solving that for 2 cases:
1. Important queries.  It does make sense to have them written in some file since we will execute them often.
2. Specific queries. We use them more rare or they are just some specific versions of general query fitting some one-time needs.
It is obvious that we have to record important queries somewhere and it does not make sense to held all variations of some general queries in all the files: it will make query list messy and will be full of code duplications. 
The problem with specific queries is that (please, correct me if I got that wrong) for any query to get executed it follows one execution path where it is being loaded from file to dict, then we use jsone.render and dict of values taken from argparse to replace variable values, finally we convert updated query to the ActiveData in json format and output the response as we desire. So currently, query is not being modified during the execution path (jsone render fits $eval values, but does not update the logic), but if we would want to have some general queries which we would want to use with only some variation in logic, we should allow query modification in query.py.
My proposal is to update couple of line in `run_query` method and add 2 methods in `query.py`:
Firstly:
```
def run_query(name, optional_arguments, config, **context):
    for query in load_query(name):
        if 'limit' in context:
            query['limit'] = context['limit']
        if 'format' in context:
            query['format'] = context['format']
        if config.debug:
            query['meta'] = {"save": True}
        **update_draft_query(query, context, optional_arguments)**
        query = jsone.render(query, context)
        query_str = json.dumps(query, indent=2, separators=(',', ':'))
        log.debug("Running query {}:\n{}".format(name, query_str))
        yield query_activedata(query_str, config.url)
```

**Comment by ahal** [comment]
Ok, good investigating! We could confirm with ekyle, but I guess that ActiveData behaves this way for exactly this reason! (So we don't have to write separate queries to handle null values)

File: adr/recipes/skipped_tests.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                        help="Select platform to get skipped tests from." +
+                        "By default, gets information about tests across all platforms.")
+    args = parser.parse_args(args)
+    query_args = vars(args)
+    result = next(
+        run_query(
+            'skipped_tests',
+            config,
+            **query_args))['data']
+    result.insert(0, ['Result test', 'run suite', 'count'])
```
Comment (ahal) [inline adr/recipes/skipped_tests.py]: It would be handy to sort the result by `path` first. You can use something like `result.sort(key=lambda x: x[0])`

**Comment by MadinaB** [comment]
Thank you!

Sorting by path added as you suggested!

Here is before:
screen shot 2018-11-14 at 2 35 51 am

Here is after:
screen shot 2018-11-14 at 2 36 18 am


Best,

**Comment by ahal** [comment]
This is amazing, thank you so much!

I have some ideas for improving this if you'd like to keep working on this recipe. But I can also understand if you'd like to focus on something else for a bit.

**Comment by MadinaB** [comment]
Thank you @ahal,

Yes, I would like to continue working on this and would be glad to get any new directions.

**Pull request merged**